### ISSUE TITLE
Redundant / multi-phase CQT basis

### ISSUE THREAD

**Issue by bmcfee** [librosa/librosa#342]
As noted in #341, the CQT basis can be lossy for high-frequency elements when a large hop-length is used.

[Schoerkhuber & Klapuri](https://www.researchgate.net/profile/Christian_Schoerkhuber/publication/228523955_Constant-Q_transform_toolbox_for_music_processing/links/55126aa60cf20bfdad513a3f.pdf) circumvent this by using a redundant basis where the same kernel is replicated several times at different phase offsets.  This gives good preservation of transients, and is generally necessary for high-quality inversion #165 .

Implementing this will require several changes to the code, basically anywhere a CQT is touched.  This includes:
- `filters.constant_q` and friends
- `filters.cq_to_chroma`
- `core.cqt`, hybrid, pseudo
- `feature.chroma_cqt` and friends
- `core.cqt_frequencies`
- `display.specshow`
- ... and possibly others

**Comment by carlthome** [comment]
First of all, the 2010 paper and [the more recent 2014 CQT Toolbox](http://www.cs.tut.fi/sgn/arg/CQT/schoerkhuber-aes-2014.pdf) are different. News are a perfect inverse and a variable Q such that bass frequencies don't require freakishly long windows. Are you familiar with the latest one and think it would be worthwhile or is the 2010 version enough to chew on?

I think I'll need to read up on the papers a bit first, and I'll get back to you.

**Comment by bmcfee** [comment]
I'd rather handle VQT as a separate issue.

**Comment by bmcfee** [comment]
Finally coming around to this issue again.

It's simple enough to implement this and make it work, but there's a question of how best to design the API.  The current CQT implementation has a number of parameters that indirectly influence the frequency basis (and shape of the output matrix):

- n_bins
- bins_per_octave
- f_min

The resulting CQT will have `n_bins` rows, where the bottom row corresponds to `f_min`, and the frequency doubles every `bins_per_octave` rows.

To implement multi-phase bases, we'll need a new parameter `redundancy [int > 0]`, which specifies how many shifted copies of each filter we retain (default=1).  If we leave all other parameters as before, then we will simply multiply everything by `redundancy`.
The output is `n_bins * redundancy`, and frequencies double every `bins_per_octave * redundancy`.  The semantics here are that a *bin* is actually multiple output rows now.

Is this confusing?

Would it be more or less confusing to assume that the user has already factored `redundancy` into the `n_bins` and `bins_per_octave` parameters?  This seems harder to me, but also more consistent with the existing relationship between `n_bins` and `bins_per_octave`.

Two more related questions:

1. Should we add redundancy symmetrically in time?  Would this require that `redundancy` is an odd integer?  Or should we only be redundant to the right (increasing in time)?  The latter ought to be sufficient.
2. K&S recommend using different per-atom hop lengths which ultimately depend on properties of the window function.  For now, since we're using hann windows everywhere, 1/4 seems like a reasonable default.  However, we might want to automate this based on similar logic to the bandwidth computation in #399.  Or should it just be a user-supplied parameter?

**Comment by bmcfee** [comment]
@carlthome Hey, I finally got around to working through the NSGT paper (Velasco 2011) and the 2014 paper for the CQT toolbox, and I think I have a grasp on what's going on in that work relative to the 2010 paper and librosa's implementation.  (Note: I have not looked into their implementations or source code.)

I wanted to jot down some thoughts here, and if I have any of this wrong,

After much unpacking, the main feature I've noticed in the 2011 paper is that hop length is inferred from the filter frequencies.  Specifically, if you have sampling rate `sr`, max frequency `fmax`, and `B` bins per octave, then they require
```
hop_length <= sr / fmax * (2**(1/B) - 2**(-1/B))
```
for reconstruction (section 3.1, paragraph 4).  (Actually, they consider the case where each frequency gets its own hop length, but for the sake of standardizing the representation, you take the maximum allowable hop corresponding to the highest frequency.)

As far as I can tell, this applies to the 2014 paper as well. The only difference in the 2014 paper is that they match the phase of the transform coefficients to the hop=1 (and then downsampled) case.

In practice, this number turns out to be tiny -- much smaller than the typical hop lengths we use elsewhere -- and you'll notice that they never actually describe the shape of the transform output in the reconstruction experiments.  For instance, if `sr=22050`, `fmax=2093` ('C7') and `B=12`, then you get `hop<=1.22`, so you're operating at the full time resolution.  In this case, it's not at all surprising that you can get perfect reconstruction.

---

Back to the topic of this issue: the NSGT method seems interesting, but I think the redundant/multi-phase implementation fits better with the rest of the package since it allows you to specify the hop length explicitly, but still compensate for undersampling.

As I see it, this comes down to a question of putting redundancy in the time domain or the frequency domain.  If it's in the frequency domain, then it's easier to align the (fixed frame rate) output to other features computed on the same time grid, and this seems like a big win to me.

Does this all seem correct and/or reasonable?

**Comment by bmcfee** [comment]
Pushing this to 0.5.1, since I don't think it will require any API changes that break compatibility.

**Comment by carlthome** [comment]
Ping @johnPertoft

**Comment by carlthome** [comment]
> you take the maximum allowable hop corresponding to the highest frequency.

Not necessarily. Their MATLAB implementation supports jagged arrays too, but yeah a "fully-rasterized" matrix by the smallest hop length is a more useful representation in most applications, I'd wager. I would not miss anything if LibROSA only supported that.

> For instance, if sr=22050, fmax=2093 ('C7') and B=12, then you get hop<=1.22, so you're operating at the full time resolution.

Ya, noticed this in @johnPertoft's NumPy implementation. @johnPertoft, thoughts?

> As I see it, this comes down to a question of putting redundancy in the time domain or the frequency domain. If it's in the frequency domain, then it's easier to align the (fixed frame rate) output to other features computed on the same time grid, and this seems like a big win to me.

Yes, totally agree. :+1:

Though I would like it to be clearly documented that for lower frequency bands there are essentially duplicated frames (in order to match the higher frequencies' frame rate). This seems like it could trip up some applications.

**Comment by carlthome** [comment]
> it allows you to specify the hop length explicitly, but still compensate for undersampling.

What about gamma? It lets you choose hop length, no?

**Comment by bmcfee** [comment]
> I would not miss anything if LibROSA only supported that.

Good. :)  Ragged time-sampling is a practical nightmare and I have no interest in implementing it.



> Though I would like it to be clearly documented that for lower frequency bands there are essentially duplicated frames (in order to match the higher frequencies' frame rate). This seems like it could trip up some applications.

The way I see it, it's better if we just say that all frequency bands are replicated `k` times, not just the high-F bands where it's strictly necessary.  It's a bit inefficient, but it simplifies the logic considerably.

**Comment by lostanlen** [comment]
NB: librosa v0.6 has introduced per-channel energy normalization (`pcen`), which will not work as intended if applied on a multi-phase CQT input.
More generally, I am worried that multi-phase discrete time-frequency representations might be difficult to document and test.
One solution would be to define is as a custom Python class. Yet, librosa has been almost free of any object-oriented code thus far, except when absolutely necessary (`LibrosaError`,  `CacheManager`,  and a few more for display). Breaking that simplicity would be a heavy decision to make.
In `scatnet` and `scattering.m` (libraries for wavelets and scattering transforms), we had to roll our own data structures, where each octave is stored in a different 2-d array. That made the code efficient but was a huge loss in transparency with respect to what librosa offers, i.e. `S` as a numpy array that can be readily passed to anything: `logamplitude`, `pcen`, `A_weighting`, `delta`, `hpss`,  `display.specshow`, etc.

TLDR: feature extraction is hard. Analysis-synthesis is hard. I have yet to find a library that does both well.

**Comment by lostanlen** [comment]
> Not sure I follow the reasoning here? If the different phase offsets are stored in different channels, it shouldn't be a big deal, right?
The reasoning is that PCEN begins by applying a temporal IIR filter to each frequency bin. If that frequency bin is smeared over multiple carriers (rows), running `scipy.signal.lfilter` on each row produces aliasing. Non ?

**Comment by bmcfee** [comment]
> If that frequency bin is smeared over multiple carriers (rows), running scipy.signal.lfilter on each row produces aliasing. Non ?

I was thinking of the different phase-offset filterbanks as being independent.  So yeah, the signal will be redundantly encoded (smeared), but this should only be a problem for reconstruction, where the basis redundancy needs to explicitly handled anyway.

Maybe you had something different in mind?

**Comment by lostanlen** [comment]
Looking at this again. You're right: the application of modulus buys us a factor of about `Q` in terms of demodulation, so the different phase channels corresponding to the same CQT frequency bin will have comparable activation magnitudes, especially after applying IIR low-pass filtering. PCEN could work out of the box.

**Comment by bmcfee** [comment]
Random thought about representation: would it make sense to encode the redundant bases along a third axis?  This would be similar to HCQT (time-frequency-harmonic), except instead of harmonic, we'd have a *delay* axis.

Pros:

- Simple to implement
- The axis logic elsewhere in the module doesn't need to change
- Representation is still CNN-friendly

Cons:

- No way to obviously distinguish it from an HCQT
- The naive implementation would be pretty slow (that's probably okay)

**Comment by bmcfee** [comment]
After discussing offline, I think we've come to the conclusion that this feature is not worth pursuing given all the representational headaches it would cause.
